src/deepsparse/transformers/pipelines/embedding_extraction.py

```diff
@@ -0,0 +1,438 @@
+            )
+
+        if args:
+            if isinstance(args, str):
+                return self.input_schema(inputs=[args[0]])
+
+            if len(args) == 1:
+                if isinstance(args[0], self.input_schema):
+                    return args[0]
+                else:
+                    return self.input_schema(inputs=args[0])
+            else:
+                return self.input_schema(inputs=args)
+        else:
+            return self.input_schema(**kwargs)
+
+    def process_inputs(self, inputs: EmbeddingExtractionInput) -> List[numpy.ndarray]:
+        """
+        Tokenizes input
+
+        :param inputs: inputs to the pipeline.
+        :return: inputs of this model processed into a list of numpy arrays that
+            can be directly passed into the forward pass of the pipeline engine
+        """
+        if isinstance(inputs.inputs, str):
+            inputs.inputs = [inputs.inputs]
+
+        tokens = self.tokenizer(
+            inputs.inputs,
+            add_special_tokens=True,
+            return_tensors="np",
+            padding=PaddingStrategy.MAX_LENGTH.value,
+            truncation=TruncationStrategy.LONGEST_FIRST.value,
+        )
+
+        # mask padding and cls_token
+        pad_masks = tokens["input_ids"] == self.tokenizer.pad_token_id
+        cls_masks = tokens["input_ids"] == self.tokenizer.cls_token_id
+
+        return self.tokens_to_engine_input(tokens), {
+            "pad_masks": pad_masks,
+            "cls_masks": cls_masks,
+        }
+
+    def engine_forward(self, engine_inputs: List[numpy.ndarray]) -> List[numpy.ndarray]:
+        """
+        Handles multithreaded engine inference
+
+        :param engine_inputs: list of numpy inputs to Pipeline engine forward pass
+        :return: result of forward pass to Pipeline engine
+        """
+
+        def _engine_forward(batch_origin: int) -> List[numpy.ndarray]:
+            # run engine
+            engine_input = engine_inputs_numpy[
+                :, batch_origin : batch_origin + self._batch_size, :
+            ]
+            engine_input = [model_input for model_input in engine_input]
+            engine_output = self.engine(engine_input)
+
+            # save results
+            engine_outputs[
+                batch_origin : batch_origin + self._batch_size
+            ] = engine_output[0]
+
+            # update tqdm
+            if progress:
+                progress.update(1)
+
+        engine_inputs_numpy = numpy.array(engine_inputs)
+
+        if engine_inputs_numpy.shape[1] % self._batch_size != 0:
+            raise ValueError(
+                f"number of engine inputs {engine_inputs_numpy.shape[1]} must "
+                f"be divisible by batch_size {self._batch_size}"
+            )
+
+        engine_outputs = [None for _ in range(engine_inputs_numpy.shape[1])]
+
+        num_batches = engine_inputs_numpy.shape[1] // self._batch_size
+        progress = (
```

Comment:
let's move progress bar out of embedding extraction, either to script where top level IR pipeline is used, or to IR pipeline

Reply:
If progress is handled on the HaystackNode level, then the node won’t be able to utilize dynamic batch. But we don’t want it implemented on the EmbeddingExtractionPipeline level either. For this reason I think we should either integrate it into EmbeddingExtractionPipeline layer or not at all, leaning towards not at all.

Reply:
agreed - not at all is better especially given we plan on using async execution